dm/syncer/syncer.go

```diff
@@ -1424,48 +1408,115 @@ func (s *Syncer) syncDML() {
 	}
 }
 
-// Run starts running for sync, we should guarantee it can rerun when paused.
-func (s *Syncer) Run(ctx context.Context) (err error) {
-	runCtx, runCancel := context.WithCancel(context.Background())
-	defer runCancel()
-	tctx := s.tctx.WithContext(runCtx)
+func (s *Syncer) waitBeforeRunExit(ctx context.Context) {
```

Comment:
can you add an unit test for this function?

Reply:
most logic is tested in current ut `TestRun` , for ungraceful stop and waitTxnend test is added in it test `checkpoint_transaction`

Reply:
You forget runCancel previously https://github.com/pingcap/tiflow/pull/4035#discussion_r785619324

currently we have no test to cover this exection path I think.

Reply:
add a ut in https://github.com/pingcap/tiflow/pull/4035/commits/a06c5e82cf6d7fc3e19e4daac5bd09c6685b66d5